### PR TITLE
Probabalistic metric space

### PR DESCRIPTION
**Pull request by redhog**
Point pairs can be sampled randomly. Two random subset of all points are chosen, and the distance matrix is calculated only between these two subsets. The size of each subset is set by `samples`: if < 1 it specifies a fraction of all points, if >= 1 it specifies the number of points in each subset.

Closes https://github.com/mmaelicke/scikit-gstat/issues/48

Only look at this _after_ merging https://github.com/mmaelicke/scikit-gstat/pull/84

### CODE REVIEW

**Comment by mmaelicke** [comment]
wow it's going really fast here...
I'll try to keep up with your pace :)

**Comment by redhog** [comment]
It's not at all as large a PR as it looks as it contains all of https://github.com/mmaelicke/scikit-gstat/pull/84 too, that's why that hads to be merged first. This is just a single class, ~100 lines of code max...

**Comment by redhog** [comment]
But yeah, this is all stuff I'm building for a single feature in our software. For a bit of background: We have built (and are using) https://github.com/emerald-geomodelling/EmeraldTriangles to manage triangulated datasets. Among those are topography, and while that usually is delivered as a raster (geotiff) and easy to sample to the TIN (which is typically build from a grid + all measurement points), sometimes it comes as an existing TIN that we just add the new (grid + measurement) points to. In that case I wanted to use kriging on topo too, not just measurements. But for such a topo (DTM) TIN, the number of points is very large, and I ran into all the things I've been addressing in these two PRs.

**Comment by redhog** [comment]
For now the scikit-gstat kriging is done outside of EmeraldTriangles, but my goal is to have a utility function in that library that uses vertice points with a non-nan value in some column as observations, kriging to all rows with nan values in the same column.

**Comment by redhog** [comment]
@mmaelicke This is ready for merging now whenever you have time :)

File: skgstat/MetricSpace.py [hunk 1]
```diff
@@ -243,3 +243,70 @@ def dists(self):
...
         return self._dists
+
+class ProbabalisticMetricSpace(MetricSpace):
+    """Like MetricSpace but samples the distance pairs only returning a
+       `samples` sized subset. `samples` can either be a fraction of
+       the total number of pairs (float < 1), or an integer count.
+    """
+    
+    def __init__(self, coords, dist_metric="euclidean", max_dist=None, samples=0.5):
+        self.coords = coords.copy()
```
Comment (mmaelicke) [inline skgstat/MetricSpace.py:254]: I would add a docstring here, as the user should be able to directly interact with this class, right?

File: skgstat/MetricSpace.py [hunk 2]
```diff
@@ -243,3 +243,70 @@ def dists(self):
...
+    
+    @property
+    def ridx(self):
+        if self._ridx is None:
+            self._ridx = np.random.choice(len(self.coords), size=self.sample_count, replace=False)
+        return self._ridx
+    
+    @property
+    def ltree(self):
+        assert self.dist_metric == "euclidean", "A coordinate tree can only be constructed for an euclidean space"
```
Comment (mmaelicke) [inline skgstat/MetricSpace.py:287]: I will turn this into a if again ;)

File: skgstat/MetricSpace.py [hunk 3]
```diff
@@ -243,3 +243,70 @@ def dists(self):
...
+        if self._ltree is None:
+            self._ltree = scipy.spatial.cKDTree(self.coords[self.lidx,:])
+        return self._ltree
+
+    @property
+    def rtree(self):
+        assert self.dist_metric == "euclidean", "A coordinate tree can only be constructed for an euclidean space"
```
Comment (mmaelicke) [inline skgstat/MetricSpace.py:294]: same here: if statement

File: skgstat/MetricSpace.py [hunk 4]
```diff
@@ -243,3 +243,70 @@ def dists(self):
...
+        if self._rtree is None:
+            self._rtree = scipy.spatial.cKDTree(self.coords[self.ridx,:])
+        return self._rtree
+            
+    @property
+    def dists(self):
```
Comment (mmaelicke) [inline skgstat/MetricSpace.py:300]: This could need a docstring

File: skgstat/MetricSpace.py [hunk 5]
```diff
@@ -243,3 +243,70 @@ def dists(self):
...
+        if self._dists is None:
+            max_dist = self.max_dist
+            if max_dist is None:
+                max_dist = 1 + np.sqrt(((self.coords.max(axis=0) - self.coords.min(axis=0))**2).sum())
```
Comment (mmaelicke) [inline skgstat/MetricSpace.py:304]: @redhog, is this still the intended behavior if the distance metric is not euklidean?

File: skgstat/Variogram.py [hunk 6]
```diff
@@ -14,10 +14,9 @@
 from skgstat import estimators, models, binning
 from skgstat import plotting
 from skgstat.util import shannon_entropy
-from .MetricSpace import MetricSpace
+from .MetricSpace import MetricSpace, MetricSpacePair, ProbabalisticMetricSpace
```
Comment (mmaelicke) [inline skgstat/Variogram.py:17]: Do we actually need the import of MetricSpacePair here? I think it is not used anywhere in this file...

**Review by mmaelicke — CHANGES_REQUESTED** [review]
@redhog, looks good to me, nice extension.
I think with two unittests we can bring the coverage back. Both could use the Variogram to instantiate a ProbabilisticMetricSpace with two different sample sizes. That should cover all the new code.
If you want me to do that, just go ahead and assign me to the PR, I think I will find some time towards the end of the week.

File: skgstat/MetricSpace.py (hunk 5 above)
Comment (redhog) [reply skgstat/MetricSpace.py:304]: Oh, good catch. This is meant to create some value that's greater than the greatest possible distance.... Maybe just set ot to float max actually...

File: skgstat/Variogram.py (hunk 6 above)
Comment (redhog) [reply skgstat/Variogram.py:17]: It is used here:

https://github.com/mmaelicke/scikit-gstat/blob/probabalistic-metric-space/skgstat/Variogram.py#L246

just so you can do `Variogram(..., samples=1234)`

**Comment by redhog** [comment]
Btw, @mmaelicke I saw you added me to the citation! Thanks! I made an account with orcid so maybe you coul update the author list on zenodo there to point to my id https://orcid.org/0000-0002-8254-1163?

File: skgstat/MetricSpace.py [hunk 7]
```diff
@@ -243,3 +255,107 @@ def dists(self):
...
         # return
         return self._dists
+
+class ProbabalisticMetricSpace(MetricSpace):
+    """Like MetricSpace but samples the distance pairs only returning a
+       `samples` sized subset. `samples` can either be a fraction of
+       the total number of pairs (float < 1), or an integer count.
+    """
+    
+    def __init__(self, coords, dist_metric="euclidean", max_dist=None, samples=0.5):
```
Comment (mmaelicke) [inline skgstat/MetricSpace.py:265]: @redhog: we need another parameter here: `seed=None`. This should be passed down to any stochastic method you use (np.random.choice` I guess) to set a `np.random.seed`. That would be great to ensure reproducible samples for unittests and, more importantly, scientific applications.

File: skgstat/MetricSpace.py [hunk 8]
```diff
@@ -243,3 +255,107 @@ def dists(self):
...
+    def sample_count(self):
+        if isinstance(self.samples, int):
+            return self.samples
+        return int(self.samples * len(self.coords))
+        
+    @property
+    def lidx(self):
+        """The sampled indices into `self.coords` for the left sample."""
+        if self._lidx is None:
+            self._lidx = np.random.choice(len(self.coords), size=self.sample_count, replace=False)
```
Comment (mmaelicke) [inline skgstat/MetricSpace.py:304]: Then, here the `self.seed` can be passed down.


BTW. numpy is suggesting to use [Generator](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.choice.html#numpy.random.Generator.choice) over `np.random.choice` today...

File: skgstat/MetricSpace.py [hunk 9]
```diff
@@ -243,3 +255,107 @@ def dists(self):
...
+        return self._lidx
+    
+    @property
+    def ridx(self):
+        """The sampled indices into `self.coords` for the right sample."""
+        if self._ridx is None:
+            self._ridx = np.random.choice(len(self.coords), size=self.sample_count, replace=False)
```
Comment (mmaelicke) [inline skgstat/MetricSpace.py:311]: same here

**Review by mmaelicke — CHANGES_REQUESTED** [review]
@redhog, thanks for adding all the stuff, I think everything is addressed. There is one more thing that came to my mind, should be an easy fix
Best

File: skgstat/MetricSpace.py (hunk 9 above)
Comment (redhog) [reply skgstat/MetricSpace.py:311]: Wait, so np.random.seed(42) does not apply "globally" to all numpy random operations after that call? That's what I did in the unittest. Maybe I've just misunderstood np.random a bit, and my code only works by accident?

Comment (mmaelicke) [reply skgstat/MetricSpace.py:311]: No, I think I was misconcepted here. The unittests work fine, so I guess that's fine. 
The reason why I like to have something like `ProbabilisticMetricSpace(..., seed=1234)` is that the seed appears as an argument to the user indicating that the class can somewhat be seeded. 
That is basically the main reason why I would like to have such an argument... 

What do you think?

Comment (redhog) [reply skgstat/MetricSpace.py:311]: Hm, maybe it _is_ better with a parameter after all, as np.random.choice is only called when a property is first evaluated so it's non-trivial to know what the seed actually is at that point?

```
s = ProbabilisticMetricSpace()
np.random.seed(42)
s.lidx, s.ridx

```

isn't a particularly good API ;)

**Review by mmaelicke — APPROVED** [review]
Thanks! Great PR, I really like it.

**Comment by mmaelicke** [comment]
You can merge whenever you want. I will add the changelog and release the new version to PyPI, this afternoon

**Pull request merged**